### PR TITLE
For #9294: Add option to clear current site data in quick settings dialog

### PR DESCRIPTION
**Pull request by jsoriano**
Add an option to clear current site data in quick settings dialog, similar to what can be found in the "Site security" menu in Firefox for desktop. Solves https://github.com/mozilla-mobile/fenix/issues/9294.

There are some open questions I would have:
* It would make sense to show this option only in sites that have cookies, but I couldn't find a way to check if current site has any cookie. Should this be exposed by GeckoView? Or is it ok by now to show this option for sites that don't have cookies?
* ~~Current PR obtains the base domain using the implementation in https://guava.dev/, it would be nice to use the same implementation Gecko uses, but I couldn't find it exposed. I guess it would be the one in [EffectiveTLDService](https://searchfox.org/mozilla-central/rev/fb8d77331582639ea6848a61dd8ee812fac31b77/netwerk/dns/nsEffectiveTLDService.cpp).~~ Solved thanks to [this comment](https://github.com/mozilla-mobile/fenix/pull/22455#discussion_r759543474).
* UX:
  * Is it ok to show this option in the quick settings dialog? It makes sense to me, and it would be coherent with Firefox for desktop, but I see this dialog may be a bit overloaded.
  * Should it have an icon as other options have?

It is by now hidden behind a feature flag, so if merged without solving all the open questions it would be available only in debug and nightly builds. I can remove the flag if the open questions are not concerning.

### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.
  - Accesibility scan passed, no warnings added by this feature.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by jsoriano** [comment]
Hey @Amejia481, sorry to ping you directly, I see you made many of the recent changes on the quick settings view, could you please help me finding a reviewer for this PR? Thanks!

**Comment by Amejia481** [comment]
👋🏽 No problem @jsoriano, thanks for the patch, I'll take a look!

File: app/build.gradle
Comment (Amejia481) [inline app/build.gradle:544]: We normally try to not include external dependencies unless it's strictly necessary, for multiple reason and as they contribute to increase the application size.

File: app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentState.kt
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentState.kt:75]: I think we could replace this call with [getOrigin](https://github.com/mozilla-mobile/android-components/blob/b7d7569909decf292214b4a9c2c1d81c3641b91f/components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt#L157)

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentState.kt:75]: We could pass directly the host as the inter API only removes by host see [clearData](https://github.com/mozilla-mobile/android-components/blob/b7d7569909decf292214b4a9c2c1d81c3641b91f/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt#L518)

Comment (jsoriano) [reply app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentState.kt:75]: Thanks for your comments!

The key part here is [`topPrivateDomain()`](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/net/InternetDomainName.html#topPrivateDomain()), that returns the part of the domain that is one level beneath the public suffix, and this is the tricky part. Most public suffixes are single strings as `.com`, `.org` and so on, but then there are country specific codes with secondary levels, as `.co.uk`. Even some services have specific public suffixes such as `.blogspot.com`.

Why do we need the part beneath the public suffix? Looking at the implementation in Firefox for desktop, you can see that in the list of cookies, all of them are domains with a single name plus the public suffix, no subdomains. If you click the clear site data button, you will see that the confirmation dialog always shows a top level domain without the subdomains.

I tried to pass the host directly to `clearData`, and it worked for websites without subdomains, but it didn't work for websites with subdomains. For example it works when browsing `https://elpais.com`, and clearing `elpais.com`, but it doesn't when browsing `https://www.theguardian.com` and clearing `www.theguardian.com`, for the second case to work, `clearData` must be called with `theguardian.com`.

I started trying to implement my own logic to get the top level domain, but I found it error prone, for example for the cases like `.co.uk` or `.blogspot.com`. Then I looked for implementations, and I couldn't find if the one in Gecko is exposed for android (I am guessing that Gecko uses something based on this [EffectiveTLDService](https://searchfox.org/mozilla-central/rev/fb8d77331582639ea6848a61dd8ee812fac31b77/netwerk/dns/nsEffectiveTLDService.cpp)). This is why I finally imported it from Guava.

I see some possible follow ups on this:
* I am missing something and there is a simpler way of finding the proper host for `clearData` :slightly_smiling_face: 
* We find some way of exposing the implementation in Gecko.
* We try to implement a `topPrivateDomain()` method based for example on the [Public Suffix list](https://publicsuffix.org/list/) maintained by the Mozilla Foundation.
* We continue using Guava till we find a better option.

File: app/build.gradle
Comment (jsoriano) [reply app/build.gradle:544]: Find in [my other comment](https://github.com/mozilla-mobile/fenix/pull/22455#discussion_r756729134) my reasoning about why I ended up including Guava.
I agree with the concerns, another one is that this part of the Guava API is beta. This was one of my reasons to add this behind a feature flag.
Let's make a decision in the other thread.

(If size is the main blocker at the end, I can compare the size of the apks with and without this dependency).

File: app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentState.kt
Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentState.kt:75]: Maybe @agi from the GeckoView team could help us to know if there is something similar exposed from Gecko to Android via GeckoView.

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentState.kt:75]: @jsoriano just to make you aware it may take some time for Agi to response as it's a holiday for him.

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentState.kt:75]: @jsoriano after chatting with @agi, he mentioned that the GeckoView team will be exposing an API for cleaning sites data by using a [GeckoSession instead of a text based URL](https://bugzilla.mozilla.org/show_bug.cgi?id=1743639) this way we won't need to use the Guava APIs, he also mentioned that in the meantime we could give a try to [runtime.storageController.clearDataFromBaseDomain](https://searchfox.org/mozilla-central/source/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/StorageController.java#135). To test it it will required some changes on AC for you to see it from Fenix, you will need to expose a new function on [AC](https://github.com/mozilla-mobile/android-components/blob/b7d7569909decf292214b4a9c2c1d81c3641b91f/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt#L518)

Comment (pocmo) [reply app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentState.kt:75]: Just as a side note, we do have `lib-publicsuffixlist` in A-C. :)

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentState.kt:75]: 😅 Awesome thanks @pocmo. 
 @jsorian In this case we could use `lib-publicsuffixlist `. We are already using [it in multiple places ](https://github.com/mozilla-mobile/fenix/search?q=publicSuffixList) in Fenix :)

Comment (jsoriano) [reply app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentState.kt]: Thanks for the hint! Changed in https://github.com/mozilla-mobile/fenix/pull/22455/commits/c6887cb79afe4880159884e0c3d1645a40b958bb. It seems to work as expected.

It would be still interesting if gecko could expose information about the current session, this would allow to show the option to clear data only if there is some data.

**Comment by gabrielluong** [comment]
Not to block this in anyway, @Amejia481 can you make sure to demo this to UX and Product and make sure they are aware of this new feature?

**Comment by Amejia481** [comment]
> Not to block this in anyway, @Amejia481 can you make sure to demo this to UX and Product and make sure they are aware of this new feature?

Thanks for taking a look :) 
I've already chatted with product on Slack, and we are waiting for UX approval :)

File: app/src/main/res/values-es/strings.xml
Comment (Amejia481) [inline app/src/main/res/values-es/strings.xml:1935]: Thanks for the translations, but we only update the default (`app/src/main/res/values/strings.xml`) string file as the other translations we get them from some [automatization that come from translation team](https://github.com/mozilla-mobile/fenix/pull/22639). Sorry but we will need to exclude this change as it could break the automatization.

Comment (jsoriano) [reply app/src/main/res/values-es/strings.xml]: No problem :slightly_smiling_face: Removed.

Comment (Amejia481) [reply app/src/main/res/values-es/strings.xml]: Thanks!

**Comment by Verdi** [comment]
This looks good for landing in Nightly behind a feature flag.

**Comment by Amejia481** [comment]
Thanks 👍🏽

**Comment by Amejia481** [comment]
@jsoriano because we have some limitation with our CI to fully run all tests, I had to created a new [PR](
https://github.com/mozilla-mobile/fenix/pull/22735), I added you as the author.

**Comment by jsoriano** [comment]
@Amejia481 I see there are some linting issues, should I push the fixes to your branch?

**Comment by Amejia481** [comment]
Yes, sure :)

**Review by Amejia481 — APPROVED** [review]
Just for the record ✅ 
The landing PR will be https://github.com/mozilla-mobile/fenix/pull/22735

**Comment by jsoriano** [comment]
Oh, I see it was already merged, thanks @Amejia481!

**Comment by Amejia481** [comment]
Yeah, thank you so much for all your hard work ❤️ !

**Pull request closed without merging**
